mesh/mesh.hpp

```diff
@@ -1272,46 +1275,47 @@ class Mesh
 
        The element indices in this structure don't need further processing,
        contrary to the ones obtained through Mesh::GetFacesElements and can
-       directly be used as Elem1 and Elem2 indices.
+       directly be used, e.g., Elem1 and Elem2 indices.
        Likewise the orientations for Elem1 and Elem2 already take into account
        special cases and can be used as is.
    */
    struct FaceInformation
    {
-      FaceLocation location; // Local, shared, boundary
-      FaceConformity conformity; // Conforming, non-conforming master or slave
+      FaceLocation elem_1_location, elem_2_location; // Local, shared, boundary
+      // elem1 and elem2 face conformity, i.e., conforming, non-conforming
+      // coarse, or non-conforming fine (or NA)
+      FaceConformity elem_1_conformity, elem_2_conformity;
```

Comment:
These two properties are complementary to each other -- if you know 1 then you know the other.

I suggest we combine them into 1 property:
```c++
   enum class CoarseSide { Conforming, Side1, Side2 };
   CoarseSide coarse_side;
```

Even better: to be more concise and prevent illegal configurations like `elem_2_location = NA` and `coarse_side = Side2`, we can combine `elem_2_location` and `coarse_side` into 1 property:
```c++
   enum class FaceConfig
   {
      Boundary, // always conforming
      LocalConforming, // both sides are local
      LocalNonconforming, // side 1 is the coarse side
      SharedConforming, // side 1 is always the local side
      SharedFineCoarse, // (fine face) side 1 is fine, side 2 is coarse
      SharedCoarseFine, // (fine face) side 1 is coarse, side 2 is fine
      SharedOther // other shared faces, e.g. coarse nonconforming or ghost that is not used
   }
```

Reply:
Actually, we need two constants for local non-conforming faces: `LocalFine` and `LocalCoarse`:
```c++
   enum class FaceConfig
   {
      Boundary, // always conforming
      LocalConforming, // both sides are local
      LocalFine, // local fine face, side 1 is coarse
      LocalCoarse, // local coarse face, side 1 is coarse
      SharedConforming, // side 1 is always the local side
      SharedFineCoarse, // (fine face) side 1 is fine, side 2 is coarse
      SharedCoarseFine, // (fine face) side 1 is coarse, side 2 is fine
      SharedOther // other shared faces, e.g. coarse nonconforming or ghost that is not used
   };
```

Reply:
I don't think this is making things more clear, we may want to ask the team's opinion.

Reply:
I think seeing, in one place, all possible valid configurations makes things clearer.

Reply:
What you are proposing is missing the point of `GetFaceInformation`, the purpose of `GetFaceInformation` is to have the information available to a program and NOT in the documentation. This is just bad from a design point of view to put information in the documentation when it can be in the code.

Reply:
The information provided by `elem_1_location`, `elem_2_location`, and others, can be provided as simple methods, based on the proposed `FaceConfig`.

Reply:
I think the idea of enumerating exactly the allowable configurations, and then providing convenience accessors based on that information is a good one. It will prevent illegal configurations (enforced by design), and be more explicit.

Reply:
I do not agree with this, even if I'm sensitive to your argument about illegal configurations, I think this proposition reproduces largely the flaws of `Mesh::FaceInfo`:
- The proposed naming reflects the internals, implementation details into a high-level interface meant to be as human readable as possible. The purpose of `GetFaceInformation` is to abstract the face data structure and implementation by proposing a mathematical/geometrical terminology represented by enum types. This proposition pushes users to understand the data structure and implementation, which is exactly what I'm trying to hide.
- The names in the enumerated type are not transparent as they conflate orthogonal semantics by creating unnecessary neologisms, e.g., `SharedOther` has absolutely no intuitive interpretation. I am completely open to find more appropriate terminology to describe face properties inspired by the scientific literature, but we should not conflate these properties into a single enum. Besides, this approach in general result in a combinatorial explosion of semantics which make it harder for users to understand a face representation.
- For similar reasons as what @samuelpmishLLNL explained [here](https://github.com/mfem/mfem/pull/2195#discussion_r724387693) using different enums makes the code easier to read and to interact with in a debugger. Besides, I could see benefits in modifying the `Mesh::FaceInformation` while debugging which is easier if the different semantics are separated in different enums.

Reply:
Agreed that `SharedOther` is opaque. I don't know what it means — are the ghost faces documented?

Reply:
The short answer, for users, about `SharedOther` is that these are faces you want to skip -- the other fields in the struct are undefined.

Reply:
Is there a longer answer for developers? 🙂 I'm asking because I don't understand this part of the code, but I think it would be valuable to know.

Reply:
The answer is that currently these include all ghost faces (as defined by the `ParNCMesh`) that are not shared nonconforming fine faces, see the code around this line: https://github.com/mfem/mfem/blob/bba2c08025c0700ead5965d27d7ed0f8564a0018/mesh/pncmesh.cpp#L1065 However, in `ParMesh`, or in this class, there is no useful information about them. They are simply an artifact of how the `Mesh::faces_info` is constructed right now. In principle, these unused face indices can be filtered out but that is probably not super important.

Reply:
I agree with Yohann that simplicity and human readability are important, but I also like the option to have enumerated possibilities for developers that want to dive in at that level.

Is it possible to keep `elem_1_conformity`, `elem_2_conformity`, etc. as primary and add a method `FaceType` or something that returns either `invalid` or the appropriate option from `FaceConfig`?